crates/runtime/src/datafusion.rs

```diff
@@ -1334,6 +1340,27 @@ impl DataFusion {
             }
         }
     }
+
+    /// Create or get a logical plan from the query
+    async fn get_or_create_logical_plan(
+        &self,
+        session: &SessionState,
+        sql: &str,
+    ) -> Result<LogicalPlan, DataFusionError> {
+        if let Some(plan) = self.cached_plans.get(sql).await {
+            tracing::trace!("using cached plan for {sql}");
+            return Ok(plan);
+        }
+
+        let plan = session.create_logical_plan(sql).await?;
+
+        tracing::trace!("caching plan for {sql}");
+        self.cached_plans
```

Comment:
This is in the critical query path - cloning a potentially large SQL string and the plan itself is a large price to pay.

Reply:
Can't we just store the pointer to the plan? Pinned Box?

Reply:
We can hash the SQL string. The problem with Arc'ing the Logical Plan is that every time we call `Query::run` we ultimately call `datafusion::dataframe::Dataframe::new` which takes the `plan` by value so we cannot avoid a clone of the plan. Binding parameters via `LogicalPlan::with_param_values` also consumes the plan.